tsdb/store.go

```diff
@@ -851,16 +851,29 @@ func (s *Store) DeleteShard(shardID uint64) error {
 
 	ss := index.SeriesIDSet()
 
-	s.walkShards(shards, func(sh *Shard) error {
+	err = s.walkShards(shards, func(sh *Shard) error {
 		index, err := sh.Index()
 		if err != nil {
-			return err
+			// Do not stop checking series because one shard failed
+			s.Logger.Error("cannot find shard index", zap.Uint64("shard_id", sh.ID()), zap.Error(err))
+			return nil
 		}
 
 		ss.Diff(index.SeriesIDSet())
 		return nil
 	})
 
+	// This should never happen, because walkShards only returns errors
+	//  from the function passed in, and the function above cannot
+	// return an error. But, for safety against new implementations, we
+	// check.
+	if err != nil {
+		s.Logger.Error("error walking shards", zap.Error(err))
+		// TODO(DSB): Should we give up on deleting the shard here, to avoid
+		// removing series from the series file that may exist in the shard
+		// whose index we could not load?
+	}
```

Comment:
I think the answer is "yes". If we continue deleting, we could lose series. Currently this can only happen during a restore, which probably fixes itself on restore completion, but maybe that changes in the future. Also, if the retention service is running then the shard will eventually be deleted by the retention service, or you will keep getting errors in your log. Since currently the only way to get an error is by running a restore, this will work at some point in the near future.

An alternative is to not remove series if we can't walk all the shards, but then we have phantom series that could impact performance. I prefer we leave the shard on disk until we can properly delete it without losing series or creating a performance issue.